contexts/WalletManagerContext.tsx

```diff
@@ -660,10 +668,52 @@ export function WalletManagerProvider({ children }: { children: ReactNode }) {
 
     fetchBalance();
 
-    rpcClient.addEventListener("utxos-changed", fetchBalance);
+    function checkIncomingUtxos(event: {
+      added: UtxoEntryReference[];
+      removed: UtxoEntryReference[];
+    }) {
+      if (event.added.length === 0) {
+        return;
+      }
+
+      const txId = event.added[0].outpoint.transactionId;
+      const network = settings?.networkId ?? NetworkType.Mainnet;
+      const explorerTxLink = explorerTxLinks[network];
+      const outgoingAmount = event.removed.reduce(
+        (acc, curr) =>
+          addresses.includes(curr.address?.toString() ?? "")
+            ? acc + curr.amount
+            : acc,
+        0n,
+      );
+      const incomingAmount = event.added.reduce(
+        (acc, curr) =>
+          addresses.includes(curr.address?.toString() ?? "")
+            ? acc + curr.amount
+            : acc,
+        0n,
+      );
+      const transferAmount = incomingAmount - outgoingAmount;
+      if (transferAmount <= 0) {
+        return;
+      }
+
+      internalToast.info(
+        `You’ve received ${sompiToKaspaString(transferAmount)} amount of KAS. Click to open on explorer`,
+        () => browser.tabs.create({ url: `${explorerTxLink}${txId}` }),
+        txId,
+      );
+    }
+
```

Comment:
```suggestion

    if (!addresses) return
```

I remember you said there is a bug about it will listen all addresses's utxos if it is empty.

Reply:
Yes it is guarded here https://github.com/forbole/kastle/blob/a7884563e31c85ce662bdec078280475e646d30e/contexts/WalletManagerContext.tsx#L652

The function is getting hard to read because you have a couple of nested function I will try to keep them at the top of the function.